src/Simulation/TargetDefinitions/Decompositions/Utils.qs

```diff
@@ -0,0 +1,181 @@
+                // Each time the controls is one shorter, until it is empty and the recursion stops.
+                Controlled R1(rest, (theta, qubit));
+            }
+        }
+    }
+
+    @EnableTestingViaName("Test.TargetDefinitions.ApplyGlobalPhaseFracWithR1Frac")
+    internal operation ApplyGlobalPhaseFracWithR1Frac(numerator : Int, power : Int) : Unit is Adj + Ctl {
+        body(...) {}
+        controlled(ctrls, ... ) {
+            let numControls =  Length(ctrls);
+            if (numControls > 0 ) {
+                // Invoke Controlled R1Frac, which will recursively call back into ApplyGlobalPhase.
+                // Each time the controls is one shorter, until it is empty and the recursion stops.
+                Controlled R1Frac(ctrls[1 .. numControls - 1], (numerator, power, ctrls[0]));
+            }
+        }
+    }
+
+    @EnableTestingViaName("Test.TargetDefinitions.MapPauli")
+    internal operation MapPauli(qubit : Qubit, from : Pauli, to : Pauli) : Unit is Adj {
+        if (from == to) {
+        }
+        elif ((from == PauliZ and to == PauliX) or (from == PauliX and to == PauliZ)) {
+            H(qubit);
+        }
+        elif (from == PauliZ and to == PauliY) {
+            H(qubit);
+            S(qubit);
+            H(qubit);
+        }
+        elif (from == PauliY and to == PauliZ) {
+            H(qubit);
+            Adjoint S(qubit);
+            H(qubit);
+        }
+        elif (from == PauliY and to == PauliX) {
+            S(qubit);
+        }
+        elif (from == PauliX and to == PauliY) {
+            Adjoint S(qubit);
+        }
+        else {
+            fail "Unsupported input";
+        }
+    }
+
+    @EnableTestingViaName("Test.TargetDefinitions.ReducedDyadicFraction")
+    internal function ReducedDyadicFraction(numerator : Int, denominatorPowerOfTwo : Int) : (Int,Int) {
+        if (numerator == 0) { return (0,0); }
+        mutable num = numerator;
+        mutable denPow = denominatorPowerOfTwo;
+        while(num % 2 == 0) {
+            set num /= 2;
+            set denPow += 1;
+        }
+        return (num,denPow);
+    }
+
+    @EnableTestingViaName("Test.TargetDefinitions.ReducedDyadicFractionPeriodic")
+    internal function ReducedDyadicFractionPeriodic(numerator : Int, denominatorPowerOfTwo : Int) : (Int,Int) {
+        let (k,n) = ReducedDyadicFraction(numerator,denominatorPowerOfTwo); // k is odd, or (k,n) are both 0
+        let period = 2*2^n; // \pi k / 2^n is 2\pi periodic, therefore k is 2 * 2^n periodic
+        let kMod = k % period; // if k was negative, we get kMod in a range [-period + 1, 0]
+        let kModPositive = kMod >= 0 ? kMod | kMod + period; // kModPositive is in the range [0, period - 1]
+        return (kModPositive, n);
+    }
+
+    @EnableTestingViaName("Test.TargetDefinitions.ArrayFromIndicesP")
+    internal function ArrayFromIndiciesP(values : Pauli[], indicies : Int[]) : Pauli[] {
```

Comment:
This looks like Microsoft.Quantum.Arrays.Subarray from the standard library; should that be pulled back into runtime as well?

Reply:
Good question! I'll defer to @cgranade .

Reply:
Oh, that's a good idea! I don't think this change is the place to do it, but I do agree that as we find ourselves reimplementing useful helper functions from the library in the core or foundation packages we should seriously consider pulling the corresponding library function in.